src/analysis/calculus/taylor.lean

```diff
@@ -0,0 +1,392 @@
+where $P_n f$ denotes the Taylor polynomial of degree $n$ and $f^{(n+1)}$ is the $n+1$-th iterated
+derivative. -/
+lemma taylor_mean_remainder_lagrange {f : ℝ → ℝ} {x x₀ : ℝ} {n : ℕ} (hx : x₀ < x)
+  (hf : cont_diff_on ℝ n f (Icc x₀ x))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (Icc x₀ x)) (Ioo x₀ x)) :
+  ∃ (x' : ℝ) (hx' : x' ∈ Ioo x₀ x), f x - taylor_within_eval f n (Icc x₀ x) x₀ x =
+  (iterated_deriv_within (n+1) f (Icc x₀ x) x') * (x - x₀)^(n+1) /(n+1)! :=
+begin
+  have gcont : continuous_on (λ (t : ℝ), (x - t) ^ (n + 1)) (Icc x₀ x) :=
+  by { refine continuous.continuous_on _, continuity },
+  have xy_ne : ∀ (y : ℝ), y ∈ Ioo x₀ x → (x - y)^n ≠ 0 :=
+  begin
+    intros y hy,
+    refine pow_ne_zero _ _,
+    rw [mem_Ioo] at hy,
+    rw sub_ne_zero,
+    exact hy.2.ne.symm,
+  end,
+  have hg' : ∀ (y : ℝ), y ∈ Ioo x₀ x → -(↑n + 1) * (x - y) ^ n ≠ 0 :=
+  λ y hy, mul_ne_zero (neg_ne_zero.mpr (nat.cast_add_one_ne_zero n)) (xy_ne y hy),
+  -- We apply the general theorem with g(t) = (x - t)^(n+1)
+  rcases taylor_mean_remainder hx hf hf' gcont (λ y _, monomial_has_deriv_aux y x _) hg'
+    with ⟨y, hy, h⟩,
+  use [y, hy],
+  simp only [sub_self, zero_pow', ne.def, nat.succ_ne_zero, not_false_iff, zero_sub, mul_neg] at h,
+  rw [h, neg_div, ←div_neg, neg_mul, neg_neg],
+  field_simp [nat.cast_add_one_ne_zero n, nat.factorial_ne_zero n, xy_ne y hy],
+  ring,
+end
+
+/-- **Taylor's theorem** with the Cauchy form of the remainder.
+
+We assume that `f` is `n+1`-times continuously differentiable in the closed set `Icc x₀ x` and
+`n+1`-times differentiable on the open set `Ioo x₀ x`. Then there exists a `x' ∈ Ioo x₀ x` such that
+$$f(x) - (P_n f)(x₀, x) = \\frac{f^{(n+1)}(x') (x - x')^n (x-x₀)}{n!},$$
+where $P_n f$ denotes the Taylor polynomial of degree $n$ and $f^{(n+1)}$ is the $n+1$-th iterated
+derivative. -/
+lemma taylor_mean_remainder_cauchy {f : ℝ → ℝ} {x x₀ : ℝ} {n : ℕ} (hx : x₀ < x)
+  (hf : cont_diff_on ℝ n f (Icc x₀ x))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (Icc x₀ x)) (Ioo x₀ x)) :
+  ∃ (x' : ℝ) (hx' : x' ∈ Ioo x₀ x), f x - taylor_within_eval f n (Icc x₀ x) x₀ x =
+  (iterated_deriv_within (n+1) f (Icc x₀ x) x') * (x - x')^n /n! * (x - x₀) :=
+begin
+  have gcont : continuous_on id (Icc x₀ x) := continuous.continuous_on (by continuity),
+  have gdiff : (∀ (x_1 : ℝ), x_1 ∈ Ioo x₀ x → has_deriv_at id
+    ((λ (t : ℝ), (1 : ℝ)) x_1) x_1) := λ _ _, has_deriv_at_id _,
+  -- We apply the general theorem with g = id
+  rcases taylor_mean_remainder hx hf hf' gcont gdiff (λ _ _, by simp) with ⟨y, hy, h⟩,
+  use [y, hy],
+  rw h,
+  field_simp [nat.factorial_ne_zero n],
+  ring,
+end
+
+/-- **Taylor's theorem** with a uniform bound of the remainder -/
+lemma taylor_mean_remainder_bound {f : ℝ → E} {a b : ℝ} {n : ℕ}
+  (hab : a ≤ b) (hf : cont_diff_on ℝ (n+1) f (Icc a b)) :
+  ∃ C : ℝ, ∀ (x : ℝ) (hx : x ∈ Icc a b),
+  ∥f x - taylor_within_eval f n (Icc a b) a x∥ ≤ C * (x - a)^(n+1) / n! :=
```

Comment:
By the way, if the conclusion is just of the form `there exists C such that ...`, one may as well ditch the 1/n! by putting it into the constant. Maybe a better formulation would be: assume that `f^(n+1)` is bounded by some constant C on `[a, b]`. Then ...
And then extract another lemma with an existential quantifier.

Reply:
I've added both versions